File: src/agents/nn/features.ts
/**
 * GameState → fixed-length Float32Array, from one player's perspective.
 *
 * Everything is seat-rotated (index 0 is always "me", rivals follow in turn
 * order) and roughly [0,1]-scaled, so one trained net serves every seat and
 * player count; absent seats and out-of-play systems read as zeros. The
 * layout is versioned by FEATURE_SIZE — change the layout, retrain the net.
 *
 * Known gap, accepted for v1: Court and Guild cards are encoded by counts
 * and suits, not identity, so the net cannot learn card-specific tactics.
 */
import { actionCard, ambitionCount, AMBITIONS, controlOf, courtCard, markerValue, openResourceSlots, SUITS } from '../../engine';
import type { GameState, ResourceType, VariantDef } from '../../engine';

const SYSTEMS = 24;
const MAX_SEATS = 4;
const PER_SYSTEM = 10;
const PER_AMBITION = 6;
const PER_SEAT = 12;
const HAND_EXTRAS = 6;
const COURT_SLOTS = 4;
const PER_COURT = 4;
const GLOBALS = 8;

export const FEATURE_SIZE =
  SYSTEMS * PER_SYSTEM +
  AMBITIONS.length * PER_AMBITION +
  MAX_SEATS * PER_SEAT +
  HAND_EXTRAS +
  COURT_SLOTS * PER_COURT +
  GLOBALS;

const RESOURCE_ORDER: ResourceType[] = ['material', 'fuel', 'weapon', 'relic', 'psionic'];

/**
 * Fill `out` (length FEATURE_SIZE, caller-owned) with the position as seen
 * by `player`. Zero-alloc on the hot path.
 */
export function extractFeatures(
  s: GameState,
  v: VariantDef,
  player: number,
  out: Float32Array,
): void {
  out.fill(0);
  let k = 0;

  // --- systems ---
  for (let i = 0; i < SYSTEMS; i++, k += PER_SYSTEM) {
    const st = s.systems[i];
    if (!st || st.outOfPlay) continue;
    let rivalMaxFresh = 0;
    let rivalShips = 0;
    for (let p = 0; p < s.players; p++) {
      if (p === player) continue;
      if (st.fresh[p] > rivalMaxFresh) rivalMaxFresh = st.fresh[p];
      rivalShips += st.fresh[p] + st.damaged[p];
    }
    let myCity = 0;
    let myStarport = 0;
    let rivalCities = 0;
    let rivalStarports = 0;
    for (const b of st.buildings) {
      if (b.player === player) {
        if (b.kind === 'city') myCity++;
        else myStarport++;
      } else if (b.kind === 'city') rivalCities++;
      else rivalStarports++;
    }
    const ctrl = controlOf(s, i);
    out[k] = st.fresh[player] / 4;
    out[k + 1] = st.damaged[player] / 4;
    out[k + 2] = myCity / 2;
    out[k + 3] = myStarport / 2;
    out[k + 4] = rivalMaxFresh / 4;
    out[k + 5] = rivalShips / 8;
    out[k + 6] = rivalCities / 2;
    out[k + 7] = rivalStarports / 2;
    out[k + 8] = ctrl === player ? 1 : 0;
    out[k + 9] = ctrl !== null && ctrl !== player ? 1 : 0;
  }

  // --- ambitions ---
  for (const ambition of AMBITIONS) {
    const mine = ambitionCount(s.playerStates[player], ambition);
    let bestRival = 0;
    for (let p = 0; p < s.players; p++) {
      if (p === player) continue;
      const c = ambitionCount(s.playerStates[p], ambition);
      if (c > bestRival) bestRival = c;
    }
    bestRival = Math.max(bestRival, s.phantom[ambition] ?? 0);
    let first = 0;
    let second = 0;
    for (const i of s.declared[ambition]) {
      const val = markerValue(v.ambitionMarkers, i, s.flipped[i]);
      first += val.first;
      second += val.second;
    }
    out[k] = mine / 6;
    out[k + 1] = bestRival / 6;
    out[k + 2] = first / 9;
    out[k + 3] = second / 6;
    out[k + 4] = s.declared[ambition].length / 3;
    out[k + 5] = mine > bestRival && mine > 0 ? 1 : 0;
    k += PER_AMBITION;
  }

  // --- seats, me first then rivals in turn order ---
  for (let seat = 0; seat < MAX_SEATS; seat++, k += PER_SEAT) {
    if (seat >= s.players) continue;
    const p = (player + seat) % s.players;
    const ps = s.playerStates[p];
    const open = openResourceSlots(ps);
    let agents = 0;
    for (const slot of s.court) agents += slot.agents[p] ?? 0;
    out[k] = ps.power / v.powerThreshold;
    out[k + 1] = ps.hand.length / 6;
    for (let r = 0; r < RESOURCE_ORDER.length; r++) {
      let n = 0;
      for (let i = 0; i < open; i++) if (ps.resources[i] === RESOURCE_ORDER[r]) n++;
      out[k + 2 + r] = n / 3;
    }
    out[k + 7] = ps.guildCards.length / 5;
    out[k + 8] = ps.captives.length / 8;
    out[k + 9] = ps.trophies.length / 8;
    out[k + 10] = agents / 10;
    out[k + 11] = s.initiative === p ? 1 : 0;
  }

  // --- my hand, beyond the size the seat block carries ---
  {
    const hand = s.playerStates[player].hand;
    let pips = 0;
    const suitCount = [0, 0, 0, 0];
    const topBySuit = [0, 0, 0, 0];
    for (const ps of s.playerStates) {
      for (const c of ps.hand) {
        const def = actionCard(c);
        const si = SUITS.indexOf(def.suit);
        if (def.number > topBySuit[si]) topBySuit[si] = def.number;
      }
    }
    let tops = 0;
    for (const c of hand) {
      const def = actionCard(c);
      pips += def.pips;
      const si = SUITS.indexOf(def.suit);
      suitCount[si]++;
      if (def.number === topBySuit[si]) tops++;
    }
    out[k] = pips / 24;
    out[k + 1] = tops / 4;
    for (let i = 0; i < 4; i++) out[k + 2 + i] = suitCount[i] / 6;
    k += HAND_EXTRAS;
  }

  // --- court row ---
  for (let slot = 0; slot < COURT_SLOTS; slot++, k += PER_COURT) {
    const c = s.court[slot];
    if (!c || c.card < 0) continue;
    let rivalMax = 0;
    for (let p = 0; p < s.players; p++) {
      if (p === player) continue;
      if (c.agents[p] > rivalMax) rivalMax = c.agents[p];
    }
    out[k] = (c.agents[player] ?? 0) / 6;
    out[k + 1] = rivalMax / 6;
    out[k + 2] = courtCard(c.card).kind === 'vox' ? 1 : 0;
    out[k + 3] = 1; // slot filled
  }

  // --- globals ---
  const chapter = Math.min(5, Math.max(1, s.chapter));
  out[k + (chapter - 1)] = 1; // chapter one-hot, 5 wide
  out[k + 5] = s.players / 4;
  out[k + 6] = s.actionDeck.length / 28;
  out[k + 7] = s.playerStates[player].hand.length > 0 ? 0 : 1; // out of cards
}

File: rust/crates/arcs-agents/src/nn/features.rs
//! `GameState` → fixed-length `[f32; FEATURE_SIZE]`, from one player's
//! perspective. Ported from `src/agents/nn/features.ts`.
//!
//! Everything is seat-rotated (index 0 is always "me", rivals follow in turn
//! order) and roughly 0-to-1 scaled, so one trained net serves every seat and
//! player count; absent seats and out-of-play systems read as zeros. The
//! layout is versioned by [`FEATURE_SIZE`] — change the layout, retrain the
//! net.
//!
//! Known gap, accepted for v1: Court and Guild cards are encoded by counts and
//! suits, not identity, so the net cannot learn card-specific tactics.
//!
//! Extraction is zero-alloc: the caller owns the buffer and the function only
//! writes into it, because the intended caller evaluates tens of thousands of
//! leaves per decision.

use arcs_engine::ambitions::{ambition_count, marker_value};
use arcs_engine::cards::{ACTION_CARD_COUNT, action_card};
use arcs_engine::court::{CourtCardKind, court_card};
use arcs_engine::map::SYSTEM_COUNT;
use arcs_engine::{
    AmbitionId, BuildingKind, GameState, MAX_SEATS, Player, ResourceType, Suit, VariantDef,
};

const SYSTEMS: usize = SYSTEM_COUNT;
const PER_SYSTEM: usize = 10;
const PER_AMBITION: usize = 6;
const PER_SEAT: usize = 12;
const HAND_EXTRAS: usize = 6;
const COURT_SLOTS: usize = 4;
const PER_COURT: usize = 4;
const GLOBALS: usize = 8;

/// Width of the feature vector, and the version of its layout. A net trained
/// against one width must be retrained if the layout moves; nothing here
/// silently reinterprets an old checkpoint.
pub const FEATURE_SIZE: usize = SYSTEMS * PER_SYSTEM
    + AmbitionId::COUNT * PER_AMBITION
    + MAX_SEATS * PER_SEAT
    + HAND_EXTRAS
    + COURT_SLOTS * PER_COURT
    + GLOBALS;

/// A caller-owned feature buffer.
pub type FeatureVec = [f32; FEATURE_SIZE];

/// The order resource slots are counted in — the TS `RESOURCE_ORDER`, which
/// is `ResourceType`'s own declaration order.
const RESOURCE_ORDER: [ResourceType; ResourceType::COUNT] = ResourceType::ALL;

/// Fill `out` with the position as seen by `player`. Zero-alloc on the hot
/// path.
pub fn extract_features(s: &GameState, v: &VariantDef, player: Player, out: &mut FeatureVec) {
    out.fill(0.0);
    let me = player.as_index();
    let players = s.players as usize;
    let mut k = 0usize;

    // --- systems ---
    for i in 0..SYSTEMS {
        let st = &s.systems[i];
        if st.out_of_play {
            k += PER_SYSTEM;
            continue;
        }
        let mut rival_max_fresh = 0u8;
        let mut rival_ships = 0u32;
        for p in 0..players {
            if p == me {
                continue;
            }
            rival_max_fresh = rival_max_fresh.max(st.fresh[p]);
            rival_ships += st.fresh[p] as u32 + st.damaged[p] as u32;
        }
        let mut my_city = 0u32;
        let mut my_starport = 0u32;
        let mut rival_cities = 0u32;
        let mut rival_starports = 0u32;
        for b in st.buildings.iter() {
            let city = b.kind() == BuildingKind::City;
            match (b.player() == player, city) {
                (true, true) => my_city += 1,
                (true, false) => my_starport += 1,
                (false, true) => rival_cities += 1,
                (false, false) => rival_starports += 1,
            }
        }
        let ctrl = s.control_of(arcs_engine::SystemId(i as u8));
        out[k] = st.fresh[me] as f32 / 4.0;
        out[k + 1] = st.damaged[me] as f32 / 4.0;
        out[k + 2] = my_city as f32 / 2.0;
        out[k + 3] = my_starport as f32 / 2.0;
        out[k + 4] = rival_max_fresh as f32 / 4.0;
        out[k + 5] = rival_ships as f32 / 8.0;
        out[k + 6] = rival_cities as f32 / 2.0;
        out[k + 7] = rival_starports as f32 / 2.0;
        out[k + 8] = if ctrl == Some(player) { 1.0 } else { 0.0 };
        out[k + 9] = if ctrl.is_some() && ctrl != Some(player) {
            1.0
        } else {
            0.0
        };
        k += PER_SYSTEM;
    }

    // --- ambitions ---
    for ambition in AmbitionId::ALL {
        let a = ambition.as_index();
        let mine = ambition_count(s.player(player), ambition);
        let mut best_rival = 0u8;
        for p in 0..players {
            if p == me {
                continue;
            }
            best_rival = best_rival.max(ambition_count(&s.player_states[p], ambition));
        }
        best_rival = best_rival.max(s.phantom[a]);
        let mut first = 0u32;
        let mut second = 0u32;
        for &i in s.declared[a].iter() {
            let val = marker_value(&v.ambition_markers, i as usize, s.flipped[i as usize]);
            first += val.first as u32;
            second += val.second as u32;
        }
        out[k] = mine as f32 / 6.0;
        out[k + 1] = best_rival as f32 / 6.0;
        out[k + 2] = first as f32 / 9.0;
        out[k + 3] = second as f32 / 6.0;
        out[k + 4] = s.declared[a].len() as f32 / 3.0;
        out[k + 5] = if mine > best_rival && mine > 0 {
            1.0
        } else {
            0.0
        };
        k += PER_AMBITION;
    }

    // --- seats, me first then rivals in turn order ---
    for seat in 0..MAX_SEATS {
        if seat >= players {
            k += PER_SEAT;
            continue;
        }
        let p = (me + seat) % players;
        let ps = &s.player_states[p];
        let open = ps.open_resource_slots();
        let mut agents = 0u32;
        for slot in s.court.iter() {
            agents += slot.agents[p] as u32;
        }
        out[k] = ps.power as f32 / v.power_threshold as f32;
        out[k + 1] = ps.hand.len() as f32 / 6.0;
        for (r, ty) in RESOURCE_ORDER.iter().enumerate() {
            let n = ps
                .resources
                .iter()
                .take(open)
                .filter(|slot| **slot == Some(*ty))
                .count();
            out[k + 2 + r] = n as f32 / 3.0;
        }
        out[k + 7] = ps.guild_cards.len() as f32 / 5.0;
        out[k + 8] = ps.captive_count() as f32 / 8.0;
        out[k + 9] = ps.trophy_count() as f32 / 8.0;
        out[k + 10] = agents as f32 / 10.0;
        out[k + 11] = if s.initiative.as_index() == p {
            1.0
        } else {
            0.0
        };
        k += PER_SEAT;
    }

    // --- my hand, beyond the size the seat block carries ---
    {
        let hand = &s.player(player).hand;
        let mut pips = 0u32;
        let mut suit_count = [0u32; Suit::COUNT];
        // deviation from features.ts: TS computes `topBySuit` by scanning
        // *every* player's hand (`for (const ps of s.playerStates)`), which
        // reads private information. That is harmless on a determinized world
        // and a leak on a raw observation, and the extractor is meant to run
        // on either. Rust reads only the observing seat's hand, so `tops`
        // counts the cards that top their suit *within my hand*. The layout
        // is unchanged; the numbers differ from TS, which is why the two
        // engines' nets are not interchangeable (FEATURE_SIZE versions the
        // layout, not the semantics — see the module docs).
        let mut top_by_suit = [0u8; Suit::COUNT];
        for &c in hand.iter() {
            let def = action_card(c);
            top_by_suit[def.suit.as_index()] = top_by_suit[def.suit.as_index()].max(def.number);
        }
        let mut tops = 0u32;
        for &c in hand.iter() {
            let def = action_card(c);
            pips += def.pips as u32;
            let si = def.suit.as_index();
            suit_count[si] += 1;
            if def.number == top_by_suit[si] {
                tops += 1;
            }
        }
        out[k] = pips as f32 / 24.0;
        out[k + 1] = tops as f32 / 4.0;
        for i in 0..Suit::COUNT {
            out[k + 2 + i] = suit_count[i] as f32 / 6.0;
        }
        k += HAND_EXTRAS;
    }

    // --- court row ---
    for slot in 0..COURT_SLOTS {
        if slot >= s.court.len() {
            k += PER_COURT;
            continue;
        }
        let c = &s.court.as_slice()[slot];
        let mut rival_max = 0u8;
        for p in 0..players {
            if p == me {
                continue;
            }
            rival_max = rival_max.max(c.agents[p]);
        }
        out[k] = c.agents[me] as f32 / 6.0;
        out[k + 1] = rival_max as f32 / 6.0;
        out[k + 2] = if court_card(c.card).kind == CourtCardKind::Vox {
            1.0
        } else {
            0.0
        };
        out[k + 3] = 1.0; // slot filled
        k += PER_COURT;
    }

    // --- globals ---
    let chapter = s.chapter.clamp(1, 5) as usize;
    out[k + chapter - 1] = 1.0; // chapter one-hot, 5 wide
    out[k + 5] = players as f32 / 4.0;
    out[k + 6] = s.action_deck.len() as f32 / ACTION_CARD_COUNT as f32;
    out[k + 7] = if s.player(player).hand.is_empty() {
        1.0
    } else {
        0.0
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn feature_size_is_the_pinned_layout() {
        // Versioned contract: a net trained at 348 inputs is only valid for
        // this layout. Bumping it is a retrain, never a remap.
        assert_eq!(FEATURE_SIZE, 348);
    }
}
